### PR TITLE
[SofaPython] GIL part 2

### PR DESCRIPTION
**Pull request by maxime-tournier**
This is a follow-up to #326, now with green tests.

TL;DR: There should be no unprotected python code in sofa except in extension code. 

Any code like this:

```c++
void my_function(...) {
   // ...
   PySomething_Something(...);
}
```

Should be GIL-protected as follows:
```c++
void my_function(...) {
   // ...
   // the lock runs until the end of scope
   sofa::simulation::PythonEnvironment::gil lock;
   PySomething_Something(...);
}
```

Or even better yet: don't use naked python calls and use the `PythonEnvironment` API instead. You may also temporarily release the GIL while doing a costly/blocking c++ operation as follows:

```c++
void my_function_that_already_owns_the_gil(...) {
   // ...
   // unlocks the GIL until the end of scope
   sofa::simulation::PythonEnvironment::no_gil unlock;
   // ...
}
``` 

I tried looking for all python calls in the codebase, but I may have missed some. 

# Changelog

- added `PythonEnvironment::gil` RAII for scoped acquisition of the GIL
- added `PythonEnvironment::no_gil` RAII for scoped release of the GIL
- GIL-protected most (all?) python calls with the above (except in extension code)
______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by maxime-tournier** [comment]
@hugtalbot ok will do. 

unfortunately I have to deal with the revert first :-/

**Comment by damienmarchal** [comment]
@maxime-tournier As I made the revert and it seems it cause you extra work... ? Could I have done thing differently to avoid that ?

**Comment by maxime-tournier** [comment]
@damienmarchal there's not really anything you could've done differently, that was just the first time I encountered this situation.

I tried re-merging my branch on a new PR and it succeeded, but the build failed since all the previous changes were reverted (so no `PythonEnvironment::gil` defined, etc).

I had to find and revert the revert commit, then push it to the PR branch to get the changes back in. It was very simple in the end but for a while it got me pretty worried :-)

**Comment by guparan** [comment]
[ci-build]

**Comment by maxime-tournier** [comment]
[ci-build]

**Comment by maxime-tournier** [comment]
build/tests seem good now

**Comment by damienmarchal** [comment]
Grr there is a conflict now.

**Comment by maxime-tournier** [comment]
awwwww :-/ I'll look into it monday.

**Comment by maxime-tournier** [comment]
fixed!

**Comment by guparan** [comment]
@damienmarchal: waiting your approval to merge (you are tagged as reviewer).

**Comment by damienmarchal** [comment]
Fine for me.

**Comment by maxime-tournier** [comment]
For those of you using SofaQtQuick: there is a `gil` branch with related fixes, which we just merged onto master.
